support/cas-server-support-webauthn-core-webflow/src/main/java/org/apereo/cas/webauthn/web/flow/WebAuthnPreparePrimaryLoginAction.java

```diff
@@ -0,0 +1,26 @@
+package org.apereo.cas.webauthn.web.flow;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import org.springframework.security.web.csrf.CsrfTokenRepository;
+import org.springframework.webflow.action.AbstractAction;
+import org.springframework.webflow.execution.Event;
+import org.springframework.webflow.execution.RequestContext;
+
+/**
+ * This is {@link WebAuthnPreparePrimaryLoginAction}.
+ *
+ * @author Jerome LELEU
+ * @since 6.5.0
+ */
+@RequiredArgsConstructor
+@Getter
+public class WebAuthnPreparePrimaryLoginAction extends AbstractAction {
+    private final CsrfTokenRepository csrfTokenRepository;
+
+    @Override
+    protected Event doExecute(final RequestContext requestContext) {
+        WebAuthnActionHelper.addCsrfTokenToFlowScope(requestContext, csrfTokenRepository);
```

Comment:
- You should rename this class to be a generic action class that inserts the CSRF token into the webflow.
- Remove the utility class, and move the method that inserts the CSRF token into this new class.
- Insert the new action class into every state that needs to have access to the CSRF token, rather then calling "addCsrfTokenToFlowScope" manually. The new action should be inserts into the relevant states likely as an entry action.

This way, you centralize the logic, remove the static utility method (which is not maintainable) and take advantage of the flow to do the job for you, instead of manually having to remember to insert something. You also remove the responsibility from WebAuthn action classes to have anything to do with CSRF (which has nothing to do with WebAuthn)

Reply:
What you propose is much better. I have updated the PR that way: the CSRF token addition is now an action created in the webflow configurer thanks to the `ConsumerExecutionAction` and added where needed.
Webauthn webflow actions no longer deal with the CSRF token mechanism.